File: src/interpreter.cc
#include <narwhal.h>

// typedef struct __ContextPrivate {
//     JSContextRef context;
// } ContextPrivate;
// 
// DESTRUCTOR(Context_finalize)
// {    
//     GET_INTERNAL(ContextPrivate *, data, object);
//     DEBUG("freeing Context=[%p]\n", data);
//     if (data) {
//         // free data->context
//         free(data);
//     }
// }
// END
// 
// JSClassRef Context_class(JSContextRef _context)
// {
//     static JSClassRef jsClass;
//     if (!jsClass)
//     {
//         JSClassDefinition definition = kJSClassDefinitionEmpty;
//         definition.finalize = Context_finalize;
// 
//         jsClass = JSClassCreate(&definition);
//     }
//     return jsClass;
// }
// 
// JSObjectRef Context_new(JSContextRef _context, JSValueRef *_exception)
// {
//     ContextPrivate *data = (ContextPrivate*)malloc(sizeof(ContextPrivate));
//     if (!data) return NULL;
//     
//     data->context = JSGlobalContextCreate(NULL);
//     
//     return JSObjectMake(_context, Context_class(_context), data);
// }

FUNCTION(INTERPRETER_JSGlobalContextCreate)
{
    JSGlobalContextRef new_context;

    if (JSGlobalContextCreateInGroup)
        new_context = JSGlobalContextCreateInGroup(JSContextGetGroup(_context), NULL);
    else
        new_context = JSGlobalContextCreate(NULL);

    return CALL(Context_new, new_context);
}
END

FUNCTION(INTERPRETER_JSContextGetGlobalObject, ARG_OBJ(context))
{
    GET_INTERNAL(ContextPrivate*, data, context);
    JSObjectRef global = JSContextGetGlobalObject(data->context);
    return global;
}
END

FUNCTION(INTERPRETER_JSEvaluateScript, ARG_OBJ(context), ARG_STR(source), ARG_OBJ(global), ARG_STR(sourceURL), ARG_INT(lineNumber))
{
    GET_INTERNAL(ContextPrivate*, data, context);
    
    // cleanup this exception handling
    HANDLE_EXCEPTION(true,true);
    JSStringRef sourceStr = JSValueToStringCopy(_context, source, _exception);
    HANDLE_EXCEPTION(true,true);
    JSStringRef sourceURLStr = JSValueToStringCopy(_context, sourceURL, _exception);
    HANDLE_EXCEPTION(true,true);
    JSValueRef result = JSEvaluateScript(
        data->context,
        sourceStr,
        global,
        sourceURLStr,
        lineNumber,
        _exception
    );
    HANDLE_EXCEPTION(true,true);
    
    JSStringRelease(sourceStr);
    JSStringRelease(sourceURLStr);
    
    return result;
}
END

NARWHAL_MODULE(interpreter)
{
    EXPORTS("JSGlobalContextCreate", JS_fn(INTERPRETER_JSGlobalContextCreate));
    EXPORTS("JSContextGetGlobalObject", JS_fn(INTERPRETER_JSContextGetGlobalObject));
    EXPORTS("JSEvaluateScript", JS_fn(INTERPRETER_JSEvaluateScript));
    EXPORTS("initializeNarwhal", JS_fn(NW_inititialize));
    
    NW_require("interpreter.js");
}
END_NARWHAL_MODULE

File: lib/interpreter.js
var exports = require("interpreter");

var FILE = require("file");

var Context = exports.Context = function() {
    var self = this;
    
    self._context = exports.JSGlobalContextCreate();
    self._context.toString = function() { return "[object Context]"; };
    self.global = exports.JSContextGetGlobalObject(self._context);
    self.global.toString = function() { return "[object Global]"; };
    
    this.eval = function(source) {
        source = source || "";
        var sourceURL = findSourceURL(source) || "<eval>";
    
        return exports.JSEvaluateScript(
            self._context,
            source,
            self.global,
            sourceURL,
            1
        );
    };
    
    self.evalFile = function(sourceURL) {
        var source = FILE.read(sourceURL, { charset : "UTF-8" });
        
        return exports.JSEvaluateScript(
            self._context,
            source,
            self.global,
            sourceURL,
            1
        );
    };
    
    self.Function = function() {
        var args = Array.prototype.slice.call(arguments);
        var body = args.pop() || "";
        var source = "(function("+args.join(",")+"){"+body+"/**/\n})";
        var sourceURL = findSourceURL(body) || "<function>";
        
        return exports.JSEvaluateScript(
            self._context,
            source,
            self.global,
            sourceURL,
            1
        );
    };
    
    return self;
};

function findSourceURL(source) {
    // based on https://bugs.webkit.org/show_bug.cgi?id=25475#c4
    var match = source.match(/.*\s*\/\/\s*@\s*sourceURL\s*=\s*(\S+)\s*/);
    if (match)
        return match[1];
    return null;
}
